### PR TITLE
Install setup requirements on-the-fly

### PR DESCRIPTION
**Pull request by yannikschaelte**
Automatically install setup requirements during setup.py based installation. Concerns cython+numpy, i.e. pip dependencies. Apt dependencies will need to be installed separately.

Solves #100

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -36,7 +36,7 @@ install:
   # https://github.com/conda-forge/blas-feedstock/issues/58
   - conda create -q -n test-env python=$TRAVIS_PYTHON_VERSION
   - conda activate test-env
-  - conda install -y -q lapack "libblas=*=*netlib" cython>=0.26 future>=0.15 "ipopt=$IPOPT_VERSION" numpy>=1.15 pkg-config>=0.29.2 setuptools>=39.0 six>=1.11
+  - conda install -y -q lapack "libblas=*=*netlib" "ipopt=$IPOPT_VERSION" pkg-config>=0.29.2 setuptools>=39.0
```
Comment (moorepants) [inline .travis.yml:39]: I'd prefer leaving this as is so we pre-install all dependencies from conda. We could have a separate test that attempts to install using pip installed dependencies.

File: setup.py [hunk 2]
```diff
@@ -20,6 +20,17 @@
 
 from setuptools import setup
 from setuptools.extension import Extension
+
+# install requirements before import
+from setuptools import dist
+SETUP_REQUIRES = [
+    "setuptools >= 39.0",
+    "wheel >= 0.26.2",
```
Comment (moorepants) [inline setup.py:28]: I'm not sure why wheel should be a requirement for building the library. Is that only required if you want to create a wheel?

We have #41, but I believe that requires building a wheel for ipopt first.

File: setup.py [hunk 3]
```diff
@@ -41,12 +52,13 @@
...
-                "future>=0.15",
-                "setuptools>=39.0",
-                "six>=1.11"
-                ]
+INSTALL_REQUIRES = [
+    "numpy>=1.15",
+    "cython>=0.26",
+    "future>=0.15",
+    "setuptools>=39.0",
+    "six>=1.11",
```
Comment (moorepants) [inline setup.py:60]: It is likely that cython and setuptools are not needed to run the library. We'd need to double check.

File: .travis.yml (hunk 1 above)
Comment (yannikschaelte) [reply .travis.yml]: reverted it!

File: setup.py (hunk 3 above)
Comment (moorepants) [reply setup.py:58]: I think Cython could be removed here, but not necessarily numpy. Shouldn't setuptools be a build dep then too? I don't see it used anywhere but `setup.py`.

Comment (yannikschaelte) [reply setup.py:58]: true ... I did not want to meddle too much with the build, but certainly possible that cython/numpy can be removed under install_requires (here I did not change those dependencies at all).

setuptools should technically be a build dep too, however I guess that would only work via pyproject.toml or setup.cfg, as setup.py is completely based on setuptools. afaik, it is usually pre-installed on python distros.

Comment (moorepants) [reply setup.py:58]: Understood. The file requires an import of setuptools itself, so it has to be pre-installed. We can leave this as is.

**Comment by moorepants** [comment]
Ok, thanks. Hopefully that helps a little downstream.

**Pull request merged**